### PR TITLE
Start gradual typing of frontend

### PR DESCRIPTION
**Pull request by aleksanb**
This commit introduces Typescript to our Webpack 2.0 flow.
Advantages include help from the compiler in weeding out logic errors in
components/main.js (i've simplified the websocket event type already),
and better compile time verification of our code.

I've not added ts to the backend (yet), as i couldn't figure out a
satisfactory build system just yet (just putting the compiled .js in
backend/dist could probably work).

I've yet to find a proper way of sharing files across the two projects,
as i initallly wanted to share the WSEvent type, so that i'd be
guaranteed the data sent over the socket matched the structure of the
one received.

### CODE REVIEW

File: nin/frontend/app/scripts/components/main.tsx [hunk 1]
```diff
@@ -183,15 +191,14 @@ class Main extends React.Component {
...
+        delete globalJSErrors[event.name];
         this.setState({globalJSErrors});
       } catch (e) {
         e.context = "WS load of " + event.name + " failed";
-        e.type = event.type;
         e.name = event.name;
 
         const globalJSErrors = Object.assign({}, this.state.globalJSErrors);
-        globalJSErrors[event.type] = e;
+        globalJSErrors[event.name] = e;
```
Comment (stianjensen) [inline nin/frontend/app/scripts/components/main.tsx]: changing from type to kind vs name here. What impact does this have on the behavior?

**Comment by sigvef** [comment]
s/initallly/initially/ (commit message)

File: nin/frontend/app/scripts/components/main.tsx (hunk 1 above)
Comment (sigvef) [inline nin/frontend/app/scripts/components/main.tsx]: Why are we changing it? And can't we just use the type system directly?

File: nin/frontend/app/scripts/socket.ts [hunk 2]
```diff
@@ -1,6 +1,15 @@
 const SockJS = require('sockjs-client');
 
-class SocketController {
+export interface WSEvent {
```
Comment (sigvef) [inline nin/frontend/app/scripts/socket.ts]: Perhaps we can call this WebSocketEvent for increased clarity?

File: nin/frontend/app/scripts/components/main.tsx (hunk 1 above)
Comment (aleksanb) [inline nin/frontend/app/scripts/components/main.tsx]: I switched back to using type as the name.

File: nin/frontend/app/scripts/socket.ts (hunk 2 above)
Comment (aleksanb) [inline nin/frontend/app/scripts/socket.ts]: Renamed.

**Comment by aleksanb** [comment]
Regarding using a union type (as currently used) versus inheritance to discriminate the websocket types,
union types with an explicit type attribute has the advantage of simplifying deserialization of events coming over the wire.
When events are sent from the backend, we either need to specify which kind of type the event is, or some clever serialization scheme for something like `class GraphEvent extends WebSocketEvent` so that we can deserialize again on the frontend.

Continuing to use the `type` attribute feels like less friction, and a better fit for the events we send.
In addition, using a sum type allows us to have different fields for the events sent in the future, as even though the field names for all variants are the same right now, they might not be so in the future.

**Comment by sigvef** [comment]
> In addition, using a sum type allows us to have different fields for the events sent in the future, as even though the field names for all variants are the same right now, they might not be so in the future.

This is true for inheritance too, so there is no difference there.

> When events are sent from the backend, we either need to specify which kind of type the event is, or some clever serialization scheme for something like class GraphEvent extends WebSocketEvent so that we can deserialize again on the frontend.

I get the feeling that somewhere down the line we're probably going to want to come up with serialization anyway as we add more websocket events, but that's more of a future issue.

In the end, the pros and cons for the two approaches aren't really that different, and we can always change it later again if need be! Now to actually test this thing and see if it works on my machine :D

**Comment by sigvef** [comment]
when running make:

image

**Comment by sigvef** [comment]
node 7.9.0, for reference

**Comment by aleksanb** [comment]
Should work now. I've had to disable uglify as it can't handle any newer than ES5 (so no native async/await in the frontend then).
In addition, there were so many indentation errors which showed up now for some reason, i've changed indentation to warn from err in eslintrc.

**Comment by stianjensen** [comment]
Is eslint running before or after typescript has altered the source?

**Comment by stianjensen** [comment]
You disabled uglify by not setting NODE_ENV to production? Make note that this also disables the production version of react: http://dev.topheman.com/make-your-react-production-minified-version-with-webpack/

**Comment by aleksanb** [comment]
I've updated the PR to make Typescript target es5.
This is required as long as we want to keep using uglify (which only handles es5 at the moment).
Uglify turns out to be really difficult to remove from webpack, and we only needed minimal tweaks to the codebase to get it to transpile to es5.

Another workaround is disabling webpack production mode, but this also disables react optimization, reducing performance again.
As soon as webpack updates to uglify 2/3 we can switch back to targeting es2017 or something like that.

**Pull request merged**